### PR TITLE
MAINT: Use with statement to open/close files to fix LGTM alerts

### LINKED ISSUE #9612: use with statement to open/close files.
Now that with statement - context managers has become a thing. Please feel free to use them whenever we need to read/write to file.

You can find the files here - https://lgtm.com/projects/g/numpy/numpy/alerts/?mode=tree&severity=warning&rule=10030095

### PR DESCRIPTION
**Pull request by bharatr21**
Closes #9612 by using `with` context manager to open/close files where applicable, also hopefully fixes LGTM alerts :smile: 

Since the diff is huge, could definitely involve some linting issues, or some extraneous `close()` statements could still be present. Kudos to the people who review this, and your patience is much appreciated!

### CODE REVIEW

**Comment by Bharat123rox** [comment]
The main cause of the error , seems to be
```
/usr/lib/<python3.5 or 3.6>/distutils/dist.py:261: UserWarning: Unknown distribution option: 'define_macros'
  warnings.warn(msg)
```
in the Azure Pipeline builds, Travis CI as well as Shippable.
How to find out whether any section of my diff was responsible or whether it's something else?
Pinging @mattip and @seberg for this

**Comment by charris** [comment]
> The main cause of the error , seems to be

It is something else, that warning has always been there. Look upward in the results and you will see a lot of actual errors. There is a bug in your code somewhere.

**Comment by seberg** [comment]
Probably something broke in the distutils part of the changes? 

A simple way to run the tests is to run:
```
python runtest.py
```
In case that somehow does not reproduce the failures, maybe try `git clean -xdf` to delete any old/non-standard files you have around (the only file you may need is the `site.cfg` maybe).

**Comment by Bharat123rox** [comment]
Running `python runtests.py` after `git clean -xdf` produced slightly more verbose error:
```
Running from numpy source directory.
/home/bharatr/anaconda3/lib/python3.7/distutils/dist.py:274: UserWarning: Unknown distribution option: 'define_macros'
  warnings.warn(msg)
error: Command "gcc -pthread -B /home/bharatr/anaconda3/compiler_compat -Wl,--sysroot=/ -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -Werror=vla -Werror=nonnull -Werror=pointer-arith -Wlogical-op -Werror=unused-function -fPIC -DNPY_INTERNAL_BUILD=1 -DHAVE_NPY_CONFIG_H=1 -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE=1 -D_LARGEFILE64_SOURCE=1 -DSCIPY_MKL_H -DHAVE_CBLAS -I/usr/local/include -I/usr/include -I/home/bharatr/anaconda3/include -Ibuild/src.linux-x86_64-3.7/numpy/core/src/umath -Ibuild/src.linux-x86_64-3.7/numpy/core/src/npymath -Ibuild/src.linux-x86_64-3.7/numpy/core/src/common -Inumpy/core/include -Ibuild/src.linux-x86_64-3.7/numpy/core/include/numpy -Inumpy/core/src/common -Inumpy/core/src -Inumpy/core -Inumpy/core/src/npymath -Inumpy/core/src/multiarray -Inumpy/core/src/umath -Inumpy/core/src/npysort -I/home/bharatr/anaconda3/include/python3.7m -Ibuild/src.linux-x86_64-3.7/numpy/core/src/common -Ibuild/src.linux-x86_64-3.7/numpy/core/src/npymath -Ibuild/src.linux-x86_64-3.7/numpy/core/src/common -Ibuild/src.linux-x86_64-3.7/numpy/core/src/npymath -c numpy/core/src/multiarray/item_selection.c -o build/temp.linux-x86_64-3.7/numpy/core/src/multiarray/item_selection.o -MMD -MF build/temp.linux-x86_64-3.7/numpy/core/src/multiarray/item_selection.o.d" failed with exit status 1

Build failed!
```

Ping @seberg

File: numpy/distutils/conv_template.py [hunk 1]
```diff
@@ -331,7 +328,10 @@ def main():
     except ValueError:
         e = get_exception()
         raise ValueError("In %s loop at %s" % (file, e))
+    finally:
+        fid.close()
     outfile.write(writestr)
+    outfile.close()
```
Comment (eric-wieser) [inline numpy/distutils/conv_template.py]: The trick here is to use:
```
try:
    file = sys.argv[1]
except IndexError:
    fid = contextlib_nullcontext(sys.stdin)
    outfile = contextlib_nullcontext(sys.stdout)
else:
    # as before
```

Then you can use a `with` statement no matter what.

Comment (eric-wieser) [inline numpy/distutils/conv_template.py]: This line as you have it is wrong as it closes `sys.stdout`!

File: numpy/distutils/from_template.py [hunk 2]
```diff
@@ -255,10 +252,11 @@ def main():
         (base, ext) = os.path.splitext(file)
         newname = base
         outfile = open(newname, 'w')
-
     allstr = fid.read()
     writestr = process_str(allstr)
     outfile.write(writestr)
+    fid.close()
+    outfile.close()
```
Comment (eric-wieser) [inline numpy/distutils/from_template.py]: Same as above, not ok as this closes stdin/stdout.

Use `contextlib_nullcontext` here too

File: numpy/distutils/line_endings.py [hunk 3]
```diff
@@ -11,19 +11,20 @@ def dos2unix(file):
...
 
-    data = open(file, "rb").read()
-    if '\0' in data:
-        print(file, "Binary!")
-        return
+    with open(file, "rb") as fp:
+        data = fp.read()
+        if '\0' in data:
+            print(file, "Binary!")
+            return
```
Comment (eric-wieser) [inline numpy/distutils/line_endings.py]: No need to indent these lines - only the lines using `fp` need to be within the `with`

File: numpy/distutils/line_endings.py [hunk 4]
```diff
@@ -44,19 +45,20 @@ def unix2dos(file):
...
-        print('unix2dos:', file)
-        with open(file, "wb") as f:
-            f.write(newdata)
-        return file
-    else:
-        print(file, 'ok')
+    with open(file, "rb") as fp:
+        data = fp.read()
+        if '\0' in data:
+            print(file, "Binary!")
```
Comment (eric-wieser) [inline numpy/distutils/line_endings.py]: As above, no need to indent

File: numpy/f2py/f2py2e.py [hunk 5]
```diff
@@ -269,8 +269,8 @@ def scaninputline(inputline):
             options["f2py_wrapper_output"] = l
         elif f == 1:
             try:
-                open(l).close()
-                files.append(l)
+                with open(l) as f:
+                    files.append(l)
```
Comment (eric-wieser) [inline numpy/f2py/f2py2e.py]: This was just fine as it was before, albeit strange.

File: numpy/f2py/rules.py [hunk 6]
```diff
@@ -1257,82 +1257,77 @@ def buildmodule(m, um):
...
-            '%% This file is auto-generated with f2py (version:%s)\n' % (f2py_version))
-        if 'shortlatex' not in options:
-            f.write(
-                '\\documentclass{article}\n\\usepackage{a4wide}\n\\begin{document}\n\\tableofcontents\n\n')
-        f.write('\n'.join(ar['latexdoc']))
-        if 'shortlatex' not in options:
-            f.write('\\end{document}')
-        f.close()
+        with open(fn, 'w') as f:
+                f.write(
```
Comment (eric-wieser) [inline numpy/f2py/rules.py]: Indent seems wrong on this line

File: numpy/ma/mrecords.py [hunk 7]
```diff
@@ -659,13 +659,12 @@ def openfile(fname):
         return fname
     # Try to open the file and guess its type
     try:
-        f = open(fname)
+        with open(fname) as f:
+            if f.readline()[:2] != "\\x":
+                f.seek(0, 0)
+                return f
```
Comment (eric-wieser) [inline numpy/ma/mrecords.py]: This changes the meaning - you can't move the with inside the try except.

There's nothing to stop you using:
```
try:
   f = open(fname)
except IoError:
    ...

# now use the with statement
with f:
   ...
```

**Review by eric-wieser — CHANGES_REQUESTED** [review]
Needs some work still

File: numpy/distutils/conv_template.py (hunk 1 above)
Comment (eric-wieser) [reply numpy/distutils/conv_template.py]: `contextlib_nullcontext` is a backport, look around for existing uses of it.

File: numpy/f2py/f2py2e.py (hunk 5 above)
Comment (Bharat123rox) [reply numpy/f2py/f2py2e.py]: IMO, this is cleaner, less strange and preserves functionality

Comment (eric-wieser) [reply numpy/f2py/f2py2e.py]: Please either undo this change, or switch to
```
with open(l) as f:
    pass
files.append(l)
```

File: numpy/ma/mrecords.py (hunk 7 above)
Comment (Bharat123rox) [reply numpy/ma/mrecords.py]: The function's intent was to return an open file handle only, so not attempting to fix the LGTM alert on this

File: numpy/distutils/line_endings.py (hunk 4 above)
Comment (Bharat123rox) [reply numpy/distutils/line_endings.py]: Done

File: numpy/distutils/line_endings.py (hunk 3 above)
Comment (Bharat123rox) [reply numpy/distutils/line_endings.py]: Done

File: numpy/distutils/conv_template.py [hunk 8]
```diff
@@ -317,21 +316,24 @@ def main():
     try:
         file = sys.argv[1]
     except IndexError:
-        fid = sys.stdin
-        outfile = sys.stdout
+        fid = contextlib_nullcontext(sys.stdin)
+        outfile = contextlib_nullcontext(sys.stdout)
```
Comment (eric-wieser) [inline numpy/distutils/conv_template.py]: Is this already imported?

Comment (Bharat123rox) [reply numpy/distutils/conv_template.py]: Nope, but how did the build pass then?

File: numpy/distutils/conv_template.py [hunk 9]
```diff
@@ -317,21 +317,24 @@ def main():
...
+    with fid:
+        allstr = fid.read()
     try:
         writestr = process_str(allstr)
     except ValueError:
         e = get_exception()
         raise ValueError("In %s loop at %s" % (file, e))
-    outfile.write(writestr)
+
+    with outfile:
```
Comment (eric-wieser) [inline numpy/distutils/conv_template.py]: Hmm, unfortunately the handling of this function still isn't correct, and without `contextlib.ExitStack` it's really hard to get right. As an example, if `open(newname)` fails then `open(file)` is not closed.

Can you revert the changes to the two functions in this patch that do this `stdin` / `stdout` juggling, and leave them to a later PR? I think everything else in this patch is trivially correct, so would be good to get in first.

File: numpy/distutils/from_template.py [hunk 10]
```diff
@@ -248,17 +249,20 @@ def main():
...
         (base, ext) = os.path.splitext(file)
         newname = base
         outfile = open(newname, 'w')
 
-    allstr = fid.read()
+    with fid:
+        allstr = fid.read()
     writestr = process_str(allstr)
-    outfile.write(writestr)
+    with outfile:
```
Comment (eric-wieser) [inline numpy/distutils/from_template.py]: This is the other incorrect one the comment above mentions

File: numpy/distutils/conv_template.py (hunk 9 above)
Comment (Bharat123rox) [reply numpy/distutils/conv_template.py]: Done

**Comment by Bharat123rox** [comment]
Gentle ping @eric-wieser to review

File: numpy/core/code_generators/generate_umath.py [hunk 11]
```diff
@@ -1154,7 +1154,6 @@ def make_code(funcdict, filename):
 
 if __name__ == "__main__":
     filename = __file__
-    fid = open('__umath_generated.c', 'w')
-    code = make_code(defdict, filename)
-    fid.write(code)
-    fid.close()
+    with open('__umath_generated.c', 'w') as fid:
+        code = make_code(defdict, filename)
```
Comment (eric-wieser) [inline numpy/core/code_generators/generate_umath.py]: Strictly this should be before the `with` statement

File: numpy/linalg/lapack_lite/fortran.py [hunk 12]
```diff
@@ -110,15 +110,14 @@ def getDependencies(filename):
...
-        m = external_pat.match(line)
-        if m:
-            names = line = line[m.end():].strip().split(',')
-            names = [n.strip().lower() for n in names]
-            names = [n for n in names if n]
-            routines.extend(names)
-    fo.close()
+    with open(filename) as fo:
+        external_pat = re.compile(r'^\s*EXTERNAL\s', re.I)
+        routines = []
```
Comment (eric-wieser) [inline numpy/linalg/lapack_lite/fortran.py]: nit: put these two lines before the `with` statement too

**Review by eric-wieser — APPROVED** [review]
Perhaps bike-shedding at this point - but try to keep anything that doesn't need the file handle out of the `with` statement

**Comment by Bharat123rox** [comment]
@eric-wieser please re-review in case review is marked as stale

**Comment by eric-wieser** [comment]
Looks great thanks - enjoy your 9 LGTM "points" :)

**Pull request merged**